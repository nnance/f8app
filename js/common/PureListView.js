/**
 * Copyright 2016 Facebook, Inc.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to
 * use, copy, modify, and distribute this software in source code or binary
 * form for use in connection with the web services and APIs provided by
 * Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use
 * of this software is subject to the Facebook Developer Principles and
 * Policies [http://developers.facebook.com/policy/]. This copyright notice
 * shall be included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE
 *
 * @flow
 */

const ListView = require('ListView');
const Dimensions = require('Dimensions');
const Platform = require('Platform');
const StyleSheet = require('StyleSheet');
const React = require('React');
const View = require('View');

type Rows = Array<Object>;
type RowsAndSections = {
  [sectionID: string]: Object;
};

export type Data = Rows | RowsAndSections;
type RenderElement = () => ?ReactElement;

type Props = {
  data: ?Data;
  renderEmptyList?: ?RenderElement;
  minContentHeight: number;
  contentInset: { top: number; bottom: number; };
};

// FIXME: Android has a bug when scrolling ListView the view insertions
// will make it go reverse. Temporary fix - pre-render more rows
const LIST_VIEW_PAGE_SIZE = Platform.OS === 'android' ? 1 : 1;

function cloneWithData(dataSource: ListView.DataSource, data: ?Data) {
  if (!data) {
    return dataSource.cloneWithRows([]);
  }
  if (Array.isArray(data)) {
    return dataSource.cloneWithRows(data);
  }
  return dataSource.cloneWithRowsAndSections(data);
}

const styles = StyleSheet.create({
  separator: {
    backgroundColor: '#eeeeee',
    height: 1,
  },
});

class PureListView extends React.Component {
  constructor(props: Props) {
    super(props);
    const dataSource = new ListView.DataSource({
      getRowData: (dataBlob, sid, rid) => dataBlob[sid][rid],
      getSectionHeaderData: (dataBlob, sid) => dataBlob[sid],
      rowHasChanged: (row1, row2) => row1 !== row2,
      sectionHeaderHasChanged: (s1, s2) => s1 !== s2,
    });

    this.state = {
      contentHeight: 0,
      dataSource: cloneWithData(dataSource, props.data),
    };

    this.renderFooter = this.renderFooter.bind(this);
    this.onContentSizeChange = this.onContentSizeChange.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.data !== nextProps.data) {
      this.setState({
        dataSource: cloneWithData(this.state.dataSource, nextProps.data),
      });
    }
  }

  onContentSizeChange(contentWidth: number, contentHeight: number) {
    if (contentHeight !== this.state.contentHeight) {
      this.setState({ contentHeight });
    }
  }
  getScrollResponder(): any {
    return this.listView.getScrollResponder();
  }

  scrollTo(...args: Array<any>) {
    this.listView.scrollTo(...args);
  }

  props: Props;

  renderFooter(): ?ReactElement {
    if (this.state.dataSource.getRowCount() === 0) {
      return this.props.renderEmptyList && this.props.renderEmptyList();
    }

    return this.props.renderFooter && this.props.renderFooter();
  }

  render() {
    const { contentInset } = this.props;
    const bottom = contentInset.bottom +
      Math.max(0, this.props.minContentHeight - this.state.contentHeight);
    return (
      <ListView
        initialListSize={1}
        pageSize={LIST_VIEW_PAGE_SIZE}
        {...this.props}
        ref={
          (node) => {
            this.listView = node;
          }
        }
        dataSource={this.state.dataSource}
        renderFooter={this.renderFooter}
        contentInset={{ bottom, top: contentInset.top }}
        onContentSizeChange={this.onContentSizeChange}
      />
    );
  }
}

PureListView.defaultProps = {
  data: [],
  contentInset: { top: 0, bottom: 0 },
  // TODO: This has to be scrollview height + fake header
  minContentHeight: Dimensions.get('window').height + 20,
  renderSeparator: (sectionID, rowID) => <View style={styles.separator} key={rowID} />,
  renderEmptyList: null,
};

module.exports = PureListView;
