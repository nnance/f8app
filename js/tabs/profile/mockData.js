export const user = {
  id: 1,
  name: 'test',
  sex: 'M',
  facebookLinked: false,
};
export const followingCount = 1800;
export const followerCount = 1800;
export const following = [{ name: 'Art Nattapat' }, { name: 'Art Art Art' }, { name: 'Art Art Art' }, { name: 'Art Art Art' }, { name: 'Art Art Art' }, { name: 'Art Art Art' }];
export const follower = [{ name: 'Art Nattapat', following: true }, { name: 'Art Art Art', following: false }, { name: 'Art Art Art', following: false }, { name: 'Art Art Art', following: true }, { name: 'Art Art Art', following: true }, { name: 'Art Art Art', following: false }];
export const myFan = [{ name: 'Art Nattapat', candys: 1250 }, { name: 'Art Art Art', candys: 1250 }, { name: 'Art Art Art', candys: 1000 }, { name: 'Art Art Art', candys: 789 }, { name: 'Art Art Art', candys: 7013 }, { name: 'Art Art Art', candys: 0 }];
export const candys = 1890;
export const activity = [{
  activity: 'like',
  title: 'แมค เดมอน ตะลุยอวกาศ',
  outline: 'The Martain กำกับโดย ริดลีย์ สก๊อต เนื้อหาเล่าถึง นักบินอวกาศที่ถูกทิ้งไว้บนดาวอังคาร',
  uri: 'http://3.bp.blogspot.com/-zW6wqY_1Me0/VYvJMOV4mcI/AAAAAAAAD-4/mB_AxhFoJH4/s1600/178491main_sig07-009-516.jpg',
  date: new Date(2016, 10, 4),
}, {
  activity: 'like',
  title: 'แมค เดมอน ตะลุยอวกาศ',
  outline: 'The Martain กำกับโดย ริดลีย์ สก๊อต เนื้อหาเล่าถึง นักบินอวกาศที่ถูกทิ้งไว้บนดาวอังคาร',
  uri: 'http://3.bp.blogspot.com/-zW6wqY_1Me0/VYvJMOV4mcI/AAAAAAAAD-4/mB_AxhFoJH4/s1600/178491main_sig07-009-516.jpg',
  date: new Date(2016, 10, 8),
}, {
  activity: 'read',
  title: 'แมค เดมอน ตะลุยอวกาศ',
  outline: 'The Martain กำกับโดย ริดลีย์ สก๊อต เนื้อหาเล่าถึง นักบินอวกาศที่ถูกทิ้งไว้บนดาวอังคาร',
  uri: 'http://www.fujisan.ne.jp/fit/th/images/ohishikouen430.jpg',
  date: new Date(2015, 10, 8),
}];

export const myClogs = [
  {
    title: 'Richy Rich! รวยมากนะ! รู้ยังคะทุกคน',
    author: {
      name: 'David Beckham',
    },
    preview: require('../../assets/A.png'),
    category: 'M',
    viewCount: 12300,
    likeCount: 1500,
    createdAt: new Date(2015, 5, 24),
  },
  {
    title: 'Money Honey คุณชายหน้าตายกับยัยขี้งก',
    author: {
      name: 'สุดสาครนอย ไทรโยค',
    },
    preview: require('../../assets/B.png'),
    category: 'N',
    viewCount: 100,
    likeCount: 55,
    createdAt: new Date(2016, 10, 8),
  },
];

export const bookmark = [
  {
    title: 'Richy Rich! รวยมากนะ! รู้ยังคะทุกคน',
    preview: require('../../assets/A.png'),
    category: 'N',
    bookmarkCount: 7,
  },
  {
    title: 'Money Honey คุณชายหน้าตายกับยัยขี้งก',
    preview: require('../../assets/B.png'),
    category: 'M',
    bookmarkCount: 3,
  },
];
