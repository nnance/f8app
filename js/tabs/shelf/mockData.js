export const topClog = {
  title: 'แมค เดมอน ตะลุยอวกาศ',
  author: 'Steve Jobs',
  review: 'The martin กำลัง บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ บลาๆๆๆๆๆ',
};

export const fakeMetaClog = [
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham',
  },
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham2',
  },
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham3',
  },
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham4',
  },
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham5',
  },
  {
    title: 'Mr.Smith Goes to somewhere',
    author: 'David Beckham6',
  },
];

export const highlightCategory = [
  {
    title: 'DRAMATIC SERIES',
    following: false,
    clogs: fakeMetaClog,
  },
  {
    title: 'NEW FUNNY CLOG',
    following: false,
    clogs: fakeMetaClog,
  },
  {
    title: 'FANTASY SCI-FI',
    following: true,
    clogs: fakeMetaClog,
  },
];
