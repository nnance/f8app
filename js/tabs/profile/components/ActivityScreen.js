import React from 'react';

import {
  Image,
  Text,
  View,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';

import moment from 'moment';

import PureListView from '../../../common/PureListView';

import NavBar from './NavBar';
import { styles as commonStyles } from '../common';
import { colors } from '../../../common/styles';

const likeIcon = require('../../../assets/profile-icon/heart.png');
const readIcon = require('../../../assets/profile-icon/read.png');

const styles = StyleSheet.create({
  rowContainer: {
    padding: 13,
  },
});

const ActivityRow = props => (<TouchableOpacity style={styles.rowContainer}>
  <View style={{ flexDirection: 'row', justifyContent: 'space-between' }}>
    <View
      style={{
        alignSelf: 'flex-start',
        flexDirection: 'row',
        alignItems: 'center',
      }}
    >
      <Image
        source={props.activity === 'like' ? likeIcon : readIcon}
        style={{
          width: 15,
          height: 15,
          resizeMode: 'contain',
        }}
      />
      <Text
        style={{
          fontSize: 13,
          color: colors.textGrey,
          marginLeft: 5,
        }}
      >
        {props.activity === 'like' ? 'liked' : 'read'} funny clog
      </Text>
    </View>
    <Text
      style={{
        alignSelf: 'flex-end',
        fontSize: 10,
        color: colors.textFadedGrey,
      }}
    >{moment(props.date).fromNow()}</Text>
  </View>
  <View style={{ paddingVertical: 10 }}>
    <Image
      style={{ height: 150, borderRadius: 5 }}
      source={{ uri: props.uri }}
    />
  </View>
  <View>
    <Text
      style={{
        fontSize: 14,
        fontWeight: 'bold',
      }}
    >
      {props.title}
    </Text>
    <Text
      style={{
        fontSize: 10,
        color: colors.textFadedGrey,
      }}
      numberOfLines={1}
    >
      {props.outline}
    </Text>
  </View>
</TouchableOpacity>);

class ActivityScreen extends React.Component {
  render() {
    return (<View style={commonStyles.listViewContainer}>
      <NavBar title="กิจกรรม" onBackPress={this.props.onBackPress} />
      <PureListView
        data={this.props.activity}
        renderRow={props => <ActivityRow {...props} />}
      />
    </View>);
  }
}

export default ActivityScreen;
export {
  ActivityRow,
};
