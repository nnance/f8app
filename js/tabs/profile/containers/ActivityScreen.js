import { connect } from 'react-redux';
import ActivityScreen from '../components/ActivityScreen';
import * as mockData from '../mockData';

export default connect(() => ({ activity: mockData.activity }))(ActivityScreen);
