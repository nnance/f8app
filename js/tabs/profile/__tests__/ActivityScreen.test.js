import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import PureListView from '../../../common/PureListView';
import ActivityScreen, { ActivityRow } from '../components/ActivityScreen';
import { activity as mockActivity } from '../mockData';

describe('ActivityScreen', () => {
  it('render pureListView with activity data', () => {
    const fakeRow = { title: 'T' };
    const wrapper = shallow(<ActivityScreen activity={mockActivity} />);
    const pureListView = wrapper.find(PureListView);
    expect(pureListView.props().data).toBe(mockActivity);
    expect(pureListView.props().renderRow(fakeRow).type).toBe(ActivityRow);
  });

  describe('ActivityRow', () => {
    it('render like activity', () => {
      const tree = renderer.create(<ActivityRow {...mockActivity[0]} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('render read activity', () => {
      const tree = renderer.create(<ActivityRow {...mockActivity[2]} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
