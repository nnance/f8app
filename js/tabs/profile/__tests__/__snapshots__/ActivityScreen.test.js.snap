exports[`ActivityScreen ActivityRow render like activity 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "padding": 13,
    }
  }
  testID={undefined}>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "flexDirection": "row",
        }
      }>
      <Image
        source={1}
        style={
          Object {
            "height": 15,
            "resizeMode": "contain",
            "width": 15,
          }
        } />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 13,
            "marginLeft": 5,
          }
        }>
        liked
         funny clog
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgba(0, 0, 0, 0.3)",
          "fontSize": 10,
        }
      }>
      5 months ago
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 10,
      }
    }>
    <Image
      source={
        Object {
          "uri": "http://3.bp.blogspot.com/-zW6wqY_1Me0/VYvJMOV4mcI/AAAAAAAAD-4/mB_AxhFoJH4/s1600/178491main_sig07-009-516.jpg",
        }
      }
      style={
        Object {
          "borderRadius": 5,
          "height": 150,
        }
      } />
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }>
      แมค เดมอน ตะลุยอวกาศ
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.3)",
          "fontSize": 10,
        }
      }>
      The Martain กำกับโดย ริดลีย์ สก๊อต เนื้อหาเล่าถึง นักบินอวกาศที่ถูกทิ้งไว้บนดาวอังคาร
    </Text>
  </View>
</View>
`;

exports[`ActivityScreen ActivityRow render read activity 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  hitSlop={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
      "padding": 13,
    }
  }
  testID={undefined}>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }>
    <View
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "flexDirection": "row",
        }
      }>
      <Image
        source={1}
        style={
          Object {
            "height": 15,
            "resizeMode": "contain",
            "width": 15,
          }
        } />
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 13,
            "marginLeft": 5,
          }
        }>
        read
         funny clog
      </Text>
    </View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "rgba(0, 0, 0, 0.3)",
          "fontSize": 10,
        }
      }>
      a year ago
    </Text>
  </View>
  <View
    style={
      Object {
        "paddingVertical": 10,
      }
    }>
    <Image
      source={
        Object {
          "uri": "http://www.fujisan.ne.jp/fit/th/images/ohishikouen430.jpg",
        }
      }
      style={
        Object {
          "borderRadius": 5,
          "height": 150,
        }
      } />
  </View>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }>
      แมค เดมอน ตะลุยอวกาศ
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      numberOfLines={1}
      style={
        Object {
          "color": "rgba(0, 0, 0, 0.3)",
          "fontSize": 10,
        }
      }>
      The Martain กำกับโดย ริดลีย์ สก๊อต เนื้อหาเล่าถึง นักบินอวกาศที่ถูกทิ้งไว้บนดาวอังคาร
    </Text>
  </View>
</View>
`;
